### ISSUE TITLE
Is Marimo a better solution than Jupyter?

### ISSUE THREAD

**Issue by mahesh-panchal** [NBISweden/workshop-reproducible-research#288]
https://marimo.io/

It solves some issues that Jupyter has.
Is it worth considering this?

**Comment by mahesh-panchal** [comment]
I also just heard about [Calkit](https://www.calkit.io) which integrates with DVC.

**Comment by johnne** [comment]
Looks very interesting. I will give marimo a try and see how to add this, probably first as Extra material for the Jupyter tutorial

**Comment by johnne** [comment]
There are several things I really like with marimo. The major downsides I see are that it only runs python code and that the integration with Quarto is not as direct. However, Quarto does have the option to [render script files](https://quarto.org/docs/computations/render-scripts.html) and since marimo notebooks are pure python (`*.py`) script files this could be a way to have Quarto render marimo notebooks as well. I'll look into this.

**Comment by johnne** [comment]
Also, the major issues with reproducibility for Jupyter, like hidden states and version control, I think are mitigated to a large extent if using Jupyter directly from VScode.

**Comment by johnne** [comment]
Nope that doesn't seem to work very well. Marimo is too much of its own ecosystem to (at the moment) easily integrate with Quarto from what I can tell. You can probably get close to a lot of the formatting that Quarto can do on jupyter notebooks with just marimo but the direct integration (via a yaml-header in jupyter for example) is not there yet.

I think we can add marimo to the extra material for the jupyter tutorial, and provide an example of the Supplementary material notebook in marimo format.

**Comment by mahesh-panchal** [comment]
Why does Marimo need to integrate with Quarto?

**Comment by johnne** [comment]
it doesn't need to, I'm just saying that currently Quarto works out of the box with Jupyter to render notebooks into a format that looks a lot better than just having a collection of notebooks that you render/export into html format. For example if you're creating a quarto project.

Since we teach Quarto in the course I would say Jupyter is, currently, a better solution because we can show how to keep using Jupyter notebooks but get some extra benefits by integrating that with Quarto.

**Comment by johnne** [comment]
Yes I had a look at that but it seemed very much a work in progress. If you get it to work with quarto please let me know.

**Comment by mahesh-panchal** [comment]
Is there something in particular you want to achieve? 
Is one of the exercise files suitable, e.g. something that's in Jypyter that works with Quarto, and I can try do the same (at some point) with Marimo?

**Comment by johnne** [comment]
I couldn't get the https://dmadisetti.github.io/quarto-marimo/ extension to work as suggested, but haven't tried extensively. From what I understand you would write a `.qmd` document with `{marimo}` chunks that quarto can render(?). But I'd like to be able to have an executable marimo document that quarto can render (if I have to do some conversion before that's ok).

The `tutorials/jupyter/supplementary_material.ipynb` file is one we use for showing how to integrate Quarto + Jupyter.

**Comment by johnne** [comment]
I don't understand if the quarto-marimo page is a working extension of if it's a placeholder for what the developer wants to be able to do in the future?

**Comment by mahesh-panchal** [comment]
I had a quick play (commands are summarized after noting what was missing).

Using the Gitpod env:
```bash
mkdir tutorials/marimo 
cd !$
ln -s ../jupyter/results ../jupyter/supplementary_material.ipynb .
pixi init
pixi project channel add bioconda
pixi add marimo quarto
pixi add --pypi altair pandas matplotlib matplotlib-inline ipython nbformat
pixi shell
```

Then to do the conversion.
```bash
marimo convert supplementary_material.ipynb -o supplementary_material.py
marimo edit supplementary_material.py
# Added stuff and saved
marimo export ipynb supplementary_material.py -o supplementary_material.quarto.ipynb
quarto render supplementary_material.quarto.ipynb
quarto render supplementary_material.quarto.ipynb --to docx
```
which produced outputs, but one needs to fiddle with the quarto yaml if one wants to make a fully self-contained outputs
